keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1080 @@
+
+```
+kind: ClaimParameter
+apiVersion: dra.e2e.example.com
+spec:
+  count: 10
+  env:
+    HELLO_WORLD: "Thanks for the fish."
+```
+
+### Notes/Constraints/Caveats
+
+For this proposal to work, kube-scheduler must have permission to read, list and
+watch arbitrary resources. This is necessary because it is not known in advance
+which CRDs will be used by DRA drivers.
+
+This may change at runtime as classes get added or removed. To avoid repeated
+GET operations for those CRDs, dynamic informers will have to be used that get
+started and stopped as needed.
+
+### Risks and Mitigations
+
+Granting kube-scheduler permission to read arbitrary types could be used by an
+attacker to extract information. However, for that an attacker first would have
+to gain control of the scheduler, get it to execute API calls, and then
+retrieve the result. This seems like a low risk.
+
+Using a single NodeResourceCapacity object per node is simple and convenient
+also for admins, but might run into size limitations. This should not be a
+problem in practice because the number of drivers per node is expected to be
+low and the capacity information from each should be small. Limits for the
+JSON fragment can be introduced to prevent excessive sizes.
+
+## Design Details
+
+### Decoding claim parameters
+
+Extracting the numeric parameters can be done without knowing the CRD:
+
+- Look up the resource based on group and kind with a REST mapper.
+
+- Retrieve the object with a `dynamic.Client` as an `unstructured.Unstructured`
+  instance (basically a nested map).
+
+- Retrieve the sub-map via the JSON path.
+
+- Re-encode that part as JSON with the normal `encode/json` package.  In the
+  example above, this will include `kind` and `apiVersion:
+  counter.dra.config.k8s.io` because those are the defaults and only allowed
+  values.
+
+- Decode into the numeric parameter with the apimachinery JSON decoder,
+  ignoring unknown fields like `env` in this example.
+
+### ResourceClass extension
+
+A new, optional field in ResourceClass enables numeric parameters for claims
+using this class:
+
+```
+type ResourceClass struct {
+    ...
+
+    // If (and only if) allocation of claims using this class is handled
+    // via numeric parameters, then NumericParameters must list all claim
+    // parameter types that are allowed for claims.
+    NumericParameters []NumericParameterType
+}
+
+type NumericParameterType struct {
+	// APIGroup is the group for the resource being referenced. It is
+	// empty for the core API. This matches the group in the APIVersion
+	// that is used when creating the resources.
+	// +optional
+	APIGroup string
+
+	// Kind is the type of resource being referenced. This is the same
+	// value as in the parameter object's metadata.
+	Kind string
+
+	// FieldPath is a dot separarated JSON path that defines where in the
+	// parameter object the numeric parameters are embedded.
+	FieldPath string
```

Comment:
This assumes a single scalar parameter to be extracted for this numerical parameter type, which seems to conflict with the statement above that the capacity request parameter may be struct (even though the example is just a single counter). Unless I misunderstand. This is where I was thinking a numerical-model specific structure that holds CEL expressions could live.

Reply:
> This assumes a single scalar parameter to be extracted for this numerical parameter type,

No, the path points to an `interface{}` in the `unstructured.Unstructured`, so the content under that path could also be an object. It's just that anything below that path then must have the structure defined by the numeric parameter type, which is less flexible than templating.

> This is where I was thinking a numerical-model specific structure that holds CEL expressions could live.

Yes, instead of FieldPath we could have a Go template or CEL expression here.